docs/development/design-risks-and-mitigations.md

```diff
@@ -0,0 +1,161 @@
+
+## Running out of memory when loading Sleeper table state.
+
+See https://github.com/gchq/sleeper/issues/4214 for the issue that tracks possible work on this.
+
+In many places in Sleeper, we need to load the current state of a table in to memory from the state store. This state
+includes the details of the partitions and the file references. Examples of places where we load this state include:
+
+- Garbage collection
+- Creation of compaction jobs
+- Metrics
+- Queries
+- The state store committer
+
+All of these currently use AWS Lambda. Currently an invocation of a lambda can have at most 10240MB of memory. Hence
+there is a risk of running out of memory if a table has lots of partitions and hence lots of file references, or if the
+state for multiple tables are loaded in to memory at the same time.
+
+Most of the places where we need to load the state only require the state of a single table to be in memory at a time.
+For example, garbage collection happens independently for each Sleeper table and so only the state of one table needs
+be in memory at a time. However, there are two places where the state of multiple tables needs to be held in memory at
+once: the query executor lambda and the state store committer lambda; these are discussed separately below. Tests have
+shown it is possible to hold the information about the file references and the partitions in a table with 132,000 leaf
+partitions and tens of files per leaf partition in memory. Thus the memory limit may not be a problem in practice.
+Moreover, the use of AWS lambda is a convenience rather than a fundamental requirement. Lambdas can be easily replaced by
+ECS services or tasks which have a memory limit of 120GB. In the cases where we have large tables, the potential
+additional cost savings due to use of on-demand lambda rather than a long-running ECS task are likely to be negligible.
+Hence the current approach appears to be viable and in future it would be relatively easy to replace the use of lambda
+with long-running services which could have more memory.
+
+The property `sleeper.statestore.provider.cache.size` controls the number of tables for which the state store
+information is held in memory at once. For the lambdas that do not need to hold the state of multiple tables in memory
+at once, setting this to 1 will mean that only one state store is in memory even if a lambda instance is reused across
+different tables. However, if we set this property to 1 for the query executor lambda then it means that if a lambda
+instance receives queries for multiple tables then it will have to reload the statestore information each time a query
+for a different table is received. This would result in slow queries. Again, this can be mitigated by replacing the
+ query executor lambda with a long running ECS task that would enable it to hold the state of mulitple tables in memory
+ at once. See the next subsection for other advantages that might bring.
```

Comment:
There's a typo at "mulitple tables", and some extra spaces at the beginning of lines.

Reply:
Fixed